File: Forms/frmFolderRefs.cs
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Data;
using System.Drawing;
using System.IO;
using System.Windows.Forms;
using Control = System.Windows.Forms.Control;
namespace eLib
    {
    public partial class frmFolderRefs
        {
        public frmFolderRefs ()
            {
            InitializeComponent ();
            }
        private void frmFolderRefs_Load (object sender, EventArgs e)
            {
            ListPathx.View = View.List;
            //in module: Public DestinationFolder As String = ""
            eLibFile.DestinationFolder = System.Windows.Forms.Application.StartupPath; // OR: Environment.SpecialFolder.Desktop
            lblPath.Text = "Select folder ...";
            Width = 1310;
            Height = 677;
            Menu2_Lectures_Click (null, null);
            }
        private void treeView1_AfterSelect (object sender, TreeViewEventArgs e)
            {
            try
                {
                //show files in the list
                ListPathx.Items.Clear ();
                string [] Files = Directory.GetFiles (treeView1.SelectedNode.FullPath, "*.*");
                lblPath.Text = treeView1.SelectedNode.FullPath;
                // Loop through them to see files
                foreach (string file in Files)
                    {
                    FileInfo fi = new FileInfo (file);
                    ListViewItem item = new ListViewItem (fi.Name);
                    switch (Strings.Right (fi.Name, 4).ToLower ())
                        {
                        case ".pdf":
                                {
                                item.ForeColor = System.Drawing.Color.Black;
                                break;
                                }
                        case ".doc":
                        case "docx":
                                {
                                item.ForeColor = System.Drawing.Color.DarkBlue;
                                break;
                                }
                        case ".ppt":
                        case "pptx":
                                {
                                item.ForeColor = System.Drawing.Color.DarkRed;
                                break;
                                }
                        case ".xls":
                        case "xlsx":
                                {
                                item.ForeColor = System.Drawing.Color.DarkGreen;
                                break;
                                }
                        case ".jpg":
                        case ".bmp":
                        case ".png":
                                {
                                item.ForeColor = System.Drawing.Color.DarkOrange;
                                break;
                                }
                        default:
                                {
                                item.ForeColor = System.Drawing.Color.DarkGray;
                                break;
                                }
                        }
                    ListPathx.Items.Add (item);
                    }
                }
            catch (Exception ex)
                {
                MessageBox.Show (ex.ToString ());
                }
            }
        private void lblPath_Click (object sender, EventArgs e)
            {
            ContextMenuStrip2.Show (Control.MousePosition);
            }
        private void blLink_Click (object sender, EventArgs e)
            {
            Menu_Assign_Click (null, null);
            }
        //Menu1
        private void Menu_Read_Click (object sender, EventArgs e)
            {
            if (ListPathx.SelectedItems.Count > 0)
                {
                Instance.Path = Strings.Trim (ListPathx.SelectedItems [0].Text);
                //MessageBox.Show (treeView1.SelectedNode.FullPath + @"\" + Instance.Path);
                long G = Interaction.Shell ("RUNDLL32.EXE URL.DLL,FileProtocolHandler " + treeView1.SelectedNode.FullPath + @"\" + Instance.Path, Microsoft.VisualBasic.Constants.vbNormalFocus);
                }
            }
        private void Menu_OpenFolder_Click (object sender, EventArgs e)
            {
            try
                {
                long G = Interaction.Shell ("RUNDLL32.EXE URL.DLL,FileProtocolHandler " + treeView1.SelectedNode.FullPath, Microsoft.VisualBasic.Constants.vbNormalFocus);
                }
            catch (Exception ex)
                {
                }
            }
        private void Menu_Assign_Click (object sender, EventArgs e)
            {
            //Do Link checked items
            // intRef = ListPaths.SelectedValue
            int tmpCNT = ListPathx.CheckedItems.Count;
            if (tmpCNT == 0)
                {
                if (ListPathx.CheckBoxes != true)
                    {
                    MessageBox.Show ("1. RightClick and switch to ListView mode\n\n2. Make sure some items are checked to be Linked", "eLib", MessageBoxButtons.OK);
                    }
                else
                    {
                    MessageBox.Show ("Make sure some items are checked to be Linked", "eLib", MessageBoxButtons.OK);
                    }
                return;
                }
            tmpCNT = 0; // use as a counter
            string strItem = "";
            Client.DialogRequestParams = 2; //1:get Project 2:get SubProject 3:get Project or SubProject
            My.MyProject.Forms.frmChooseProject.ShowDialog ();
            if (Client.DialogRequestParams == 32) //32: A SubProjects is selected from dialog //intProd=id of the selected SubProject
                {
                if (SubProject.Id < 1)
                    return;
                for (int i = 0, loopTo = ListPathx.Items.Count - 1; i <= loopTo; i++)
                    {
                    if (ListPathx.Items [i].Checked == true)
                        {
                        strItem = ListPathx.Items [i].ToString ();
                        int intRefId = FindRefId (strItem);
                        if (intRefId > 0)
                            {
                            int k;
                            k = DoAssignRefToSubProjects (intRefId, SubProject.Id);
                            if (k == 1)
                                {
                                lblPath.Text = "Assigned:  " + strItem;
                                tmpCNT = tmpCNT + 1;
                                }
                            }
                        }
                    }
                lblPath.Text = "Assigned Refs:  " + tmpCNT.ToString ();
                }
            }
        private void Menu_CopyTitle_Click (object sender, EventArgs e)
            {
            if (ListPathx.SelectedItems.Count > 0)
                {
                string tmpx = ListPathx.Text;
                if (Strings.Mid (tmpx, (tmpx.Length - 3), 1) == ".")
                    {
                    My.MyProject.Computer.Clipboard.SetText (Strings.Left (tmpx, (tmpx.Length - 4)));
                    }
                else if (Strings.Mid (tmpx, (tmpx.Length - 4), 1) == ".")
                    {
                    My.MyProject.Computer.Clipboard.SetText (Strings.Left (tmpx, (tmpx.Length - 5)));
                    }
                else
                    {
                    My.MyProject.Computer.Clipboard.SetText (tmpx);
                    }
                }
            }
        private void Menu_Inverse_Click (object sender, EventArgs e)
            {
            for (int i = 0, loopTo = ListPathx.Items.Count - 1; i <= loopTo; i++)
                {
                if (ListPathx.Items [i].Checked == true)
                    {
                    ListPathx.Items [i].Checked = false;
                    }
                else
                    {
                    ListPathx.Items [i].Checked = true;
                    }
                }
            }
        private void Menu_None_Click (object sender, EventArgs e)
            {
            for (int i = 0, loopTo = ListPathx.Items.Count - 1; i <= loopTo; i++)
                ListPathx.Items [i].Checked = false;
            }
        private void Menu_ViewList_Click (object sender, EventArgs e)
            {
            ToggleView ("viewList");
            }
        private void Menu_ViewTile_Click (object sender, EventArgs e)
            {
            ToggleView ("viewTile");
            }
        //Menu2
        private void Menu2_Papers_Click (object sender, EventArgs e)
            {
            eLibFile.DestinationFolder = User.FolderPapers;
            treeView1.Nodes.Clear ();
            DirectoryInfo di = new DirectoryInfo (eLibFile.DestinationFolder);
            TreeNode tds = treeView1.Nodes.Add (di.FullName);
            tds.Tag = di.FullName;
            tds.StateImageIndex = 0;
            LoadSubDirectories (eLibFile.DestinationFolder, tds);
            treeView1.ExpandAll ();
            treeView1.SelectedNode = treeView1.Nodes [0];
            treeView1_AfterSelect (null, null);
            }
        private void Menu2_Books_Click (object sender, EventArgs e)
            {
            eLibFile.DestinationFolder = User.FolderBooks;
            treeView1.Nodes.Clear ();
            DirectoryInfo di = new DirectoryInfo (eLibFile.DestinationFolder);
            TreeNode tds = treeView1.Nodes.Add (di.FullName);
            tds.Tag = di.FullName;
            tds.StateImageIndex = 0;
            LoadSubDirectories (eLibFile.DestinationFolder, tds);
            treeView1.ExpandAll ();
            treeView1.SelectedNode = treeView1.Nodes [0];
            treeView1_AfterSelect (null, null);
            }
        private void Menu2_Manuals_Click (object sender, EventArgs e)
            {
            eLibFile.DestinationFolder = User.FolderManuals;
            treeView1.Nodes.Clear ();
            DirectoryInfo di = new DirectoryInfo (eLibFile.DestinationFolder);
            TreeNode tds = treeView1.Nodes.Add (di.FullName);
            tds.Tag = di.FullName;
            tds.StateImageIndex = 0;
            LoadSubDirectories (eLibFile.DestinationFolder, tds);
            treeView1.ExpandAll ();
            treeView1.SelectedNode = treeView1.Nodes [0];
            treeView1_AfterSelect (null, null);
            }
        private void Menu2_Lectures_Click (object sender, EventArgs e)
            {
            eLibFile.DestinationFolder = User.FolderLectures;
            treeView1.Nodes.Clear ();
            DirectoryInfo di = new DirectoryInfo (eLibFile.DestinationFolder);
            TreeNode tds = treeView1.Nodes.Add (di.FullName);
            tds.Tag = di.FullName;
            tds.StateImageIndex = 0;
            LoadSubDirectories (eLibFile.DestinationFolder, tds);
            treeView1.ExpandAll ();
            treeView1.SelectedNode = treeView1.Nodes [0];
            treeView1_AfterSelect (null, null);
            }
        private void Menu2_SelectFolderromDialog_Click (object sender, EventArgs e)
            {
            SelectDirectory ();
            }
        //methods
        private void LoadSubDirectories (string dir, TreeNode td)
            {
            //get all subdirectories
            string [] subdirectoryEntries = Directory.GetDirectories (dir);
            foreach (string subdirectory in subdirectoryEntries)
                {
                //loop through them to see if they have any other subdirectories
                DirectoryInfo di = new DirectoryInfo (subdirectory);
                TreeNode tds = td.Nodes.Add (di.Name);
                tds.StateImageIndex = 0;
                tds.Tag = di.FullName;
                LoadSubDirectories (subdirectory, tds);
                }
            }
        private int FindRefId (string strTitle)
            {
            int FindRefIdRet = default;
            //Find ID of a Ref in tblPapers
            Db.DS.Tables ["tblRefs"].Clear ();
            string Fltr = "PaperName='" + strTitle + "'";
            using (var CnnSS = new Microsoft.Data.SqlClient.SqlConnection (Db.CnnString))
                {
                CnnSS.Open ();
                Db.DASS = new Microsoft.Data.SqlClient.SqlDataAdapter ("SELECT Distinct Papers.ID, PaperName, IsPaper, IsBook, IsManual, IsLecture FROM Links RIGHT JOIN Papers ON Links.Paper_ID = Papers.ID  WHERE (" + Fltr + ") ORDER BY Papers.ID DESC;", CnnSS);
                Db.DASS.Fill (Db.DS, "tblRefs");
                CnnSS.Close ();
                }
            FindRefIdRet = Conversions.ToInteger (Db.DS.Tables ["tblRefs"].Rows [0] [0]);
            return FindRefIdRet;
            }
        private int DoAssignRefToSubProjects (int iRef, int iProd)
            {
            int DoAssignRefToSubProjectRet = default;
            Db.strSQL = "INSERT INTO Links (Paper_ID, SubProject_ID) VALUES (@paperid, @SubProjectid)";
            DoAssignRefToSubProjectRet = 1;
            using (var CnnSS = new Microsoft.Data.SqlClient.SqlConnection (Db.CnnString))
                {
                CnnSS.Open ();
                var cmdx = new Microsoft.Data.SqlClient.SqlCommand (Db.strSQL, CnnSS);
                cmdx.CommandType = CommandType.Text;
                cmdx.Parameters.AddWithValue ("@paperid", iRef.ToString ());
                cmdx.Parameters.AddWithValue ("@SubProjectid", iProd.ToString ());
                try
                    {
                    int ix = cmdx.ExecuteNonQuery ();
                    }
                catch (Exception ex)
                    {
                    DoAssignRefToSubProjectRet = 0;
                    MessageBox.Show (ex.ToString ());
                    }
                CnnSS.Close ();
                }

            return DoAssignRefToSubProjectRet;
            }
        public void SelectDirectory ()
            {
            FolderBrowserDialog1.SelectedPath = eLibFile.DestinationFolder;   // OR Environment.SpecialFolder.Desktop
            if (FolderBrowserDialog1.ShowDialog () == DialogResult.OK)
                {
                eLibFile.DestinationFolder = FolderBrowserDialog1.SelectedPath + @"\";
                //ScanFolder (eLibFile.DestinationFolder);
                treeView1.Nodes.Clear ();
                DirectoryInfo di = new DirectoryInfo (eLibFile.DestinationFolder);
                TreeNode tds = treeView1.Nodes.Add (di.FullName);
                tds.Tag = di.FullName;
                tds.StateImageIndex = 0;
                LoadSubDirectories (eLibFile.DestinationFolder, tds);
                treeView1.ExpandAll ();
                treeView1.SelectedNode = treeView1.Nodes [0];
                treeView1_AfterSelect (null, null);
                }
            else
                {
                return;
                }
            }
        private void ToggleView (string viewMode)
            {
            switch (viewMode)
                {
                case "viewList":
                        {
                        ListPathx.View = View.List;
                        ListPathx.CheckBoxes = true;
                        btnTileList.Text = "Tile";
                        break;
                        }
                case "viewTile":
                        {
                        ListPathx.CheckBoxes = false;
                        ListPathx.View = View.Tile;
                        ListPathx.TileSize = new Size (340, 60);
                        btnTileList.Text = "List";
                        break;
                        }
                case "viewLargeIcon":
                        {
                        ListPathx.View = View.LargeIcon;
                        ListPathx.CheckBoxes = true;
                        ListPathx.TileSize = new Size (340, 120);
                        btnTileList.Text = "List";
                        break;
                        }
                default:
                        {
                        ListPathx.View = View.List;
                        ListPathx.CheckBoxes = true;
                        btnTileList.Text = "List";
                        break;
                        }
                }
            try
                {
                treeView1_AfterSelect (null, null);
                }
            catch (Exception ex)
                {
                }
            }
        //Exit
        private void Menu_Exit_Click (object sender, EventArgs e)
            {
            Dispose ();
            }
        private void lblRead_Click (object sender, EventArgs e)
            {
            Menu_Read_Click (null, null);
            }
        private void lblBack_Click (object sender, EventArgs e)
            {
            Dispose ();
            }

        private void btnBrowse_Click (object sender, EventArgs e)
            {
            Menu2_SelectFolderromDialog_Click (null, null);
            }
        private void btnTileList_Click (object sender, EventArgs e)
            {
            if (ListPathx.View == View.List)
                {
                ToggleView ("viewTile");
                }
            else
                {
                ToggleView ("viewList");
                }
            }
        }
    }

File: Forms/frmFolderRefs.vb
Public Class frmFolderRefs
    Private Sub frmFolderRefs_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        '//in module: Public DestinationFolder As String = ""
        Menu_SubFolders.Checked = False
        DestinationFolder = Application.StartupPath 'OR: Environment.SpecialFolder.Desktop
        lblPath.Text = "-"
    End Sub
    Private Sub Menu_SelectFolder_Click(sender As Object, e As EventArgs) Handles Menu_SelectFolder.Click
        FolderBrowserDialog1.SelectedPath = DestinationFolder   'OR Environment.SpecialFolder.Desktop
        If (FolderBrowserDialog1.ShowDialog() = DialogResult.OK) Then
            DestinationFolder = FolderBrowserDialog1.SelectedPath & "\"
            ScanFolder(DestinationFolder)
        Else
            Exit Sub
        End If
    End Sub
    Private Sub Menu_SubFolders_Click(sender As Object, e As EventArgs) Handles Menu_SubFolders.Click
        Select Case Menu_SubFolders.Checked
            Case True : Menu_SubFolders.Checked = False
            Case False : Menu_SubFolders.Checked = True
        End Select
        'refresh the List
        ScanFolder(DestinationFolder)
    End Sub
    Public Sub ScanFolder(strFolderName As String)
        ListPaths.Items.Clear()
        lblPath.Text = DestinationFolder
        Dim boolSubFolder As Boolean = Menu_SubFolders.Checked
        '//EXTRACT FileNames : DIR (/s) >> text files
        Select Case boolSubFolder
            Case True
                Dim tmpArg As String = "/C Dir """ & strFolderName & """ /s > " & Application.StartupPath & "eLibF.txt"
                Process.Start("cmd.exe", tmpArg).WaitForExit()
            Case False
                Dim tmpArg As String = "/C Dir """ & strFolderName & """ > " & Application.StartupPath & "eLibF.txt"
                Process.Start("cmd.exe", tmpArg).WaitForExit()
                'Process.Start("cmd.exe", "/C Dir '" & strFolderName & "' > " & Application.StartupPath & "eLibF.txt").WaitForExit()
        End Select
        '//Extract lines of interest out of textfiles!
        Dim strLine As String = ""
        Dim strName As String = ""
        Try
            FileOpen(1, Application.StartupPath & "eLibF.txt", OpenMode.Input)
            'strLine = LineInput(1)
            'strLine = LineInput(1)
            While Not EOF(1)
                strLine = Trim(LineInput(1))
                If Len(strLine) < 40 Then GoTo Lblx
                If (strLine = "") Or (InStr(1, strLine, "<DIR>") <> 0) Or (InStr(1, strLine, "Directory Of") <> 0) Or (InStr(1, strLine, "File(s) ") <> 0) Or (InStr(1, strLine, "Total Files Listed:") <> 0) Or (InStr(1, strLine, "Dir(s) ") <> 0) Then GoTo Lblx
                If InStr(1, strLine, "\") <> 0 Then GoTo Lblx
                strName = Mid(strLine, 40, Len(strLine) - 39)
                strName = RemoveExtension(strName)
                If Len(strName) > 8 Then
                    ListPaths.Items.Add(strName)
                End If
Lblx:
            End While
            FileClose(1)
        Catch ex As Exception
            FileClose(1)
            MsgBox("//Error//:   " & vbCrLf & strName & vbCrLf & ex.ToString)
        End Try
        '//Delete temporary files
        If Dir(Application.StartupPath & "eLibF.txt") <> "" Then My.Computer.FileSystem.DeleteFile(Application.StartupPath & "eLibF.txt", FileIO.UIOption.OnlyErrorDialogs, FileIO.RecycleOption.SendToRecycleBin, FileIO.UICancelOption.ThrowException)
        If Dir(Application.StartupPath & "eLibFx.txt") <> "" Then My.Computer.FileSystem.DeleteFile(Application.StartupPath & "eLibFx.txt", FileIO.UIOption.OnlyErrorDialogs, FileIO.RecycleOption.SendToRecycleBin, FileIO.UICancelOption.ThrowException)
    End Sub
    Private Sub Menu_Inverse_Click(sender As Object, e As EventArgs) Handles Menu_Inverse.Click
        For i As Integer = 0 To ListPaths.Items.Count - 1
            If ListPaths.GetItemChecked(i) = True Then
                ListPaths.SetItemChecked(i, False)
            Else
                ListPaths.SetItemChecked(i, True)
            End If
        Next i
    End Sub
    Private Sub Menu_None_Click(sender As Object, e As EventArgs) Handles Menu_None.Click
        For i As Integer = 0 To ListPaths.Items.Count - 1
            ListPaths.SetItemChecked(i, False)
        Next i
    End Sub
    Private Sub Menu_CopyTitle_Click(sender As Object, e As EventArgs) Handles Menu_CopyTitle.Click
        If ListPaths.SelectedIndex >= 0 Then
            My.Computer.Clipboard.SetText(ListPaths.Text)
        End If
    End Sub
    Private Sub Menu_Read_Click(sender As Object, e As EventArgs) Handles Menu_Read.Click
        If ListPaths.SelectedIndex >= 0 Then
            strRef = Trim(ListPaths.Text)
            If strRef <> "" Then frmReadRef.ShowDialog()
        End If
    End Sub
    Private Sub Menu_Assign_Click(sender As Object, e As EventArgs) Handles Menu_Assign.Click
        '//Do Assign checked Items
        'intRef = ListPaths.SelectedValue
        Dim tmpCNT As Integer = ListPaths.CheckedItems.Count
        If tmpCNT = 0 Then
            MsgBox("No item is selected!", vbOKOnly, "eLib")
            Exit Sub
        End If
        tmpCNT = 0 'use as a counter
        Dim strItem As String = ""
        frmChooseProject.ShowDialog()
        If Retval1 = 2 Then '//2: A Product is selected from dialog //intProd=id of the selected Product
            If intProd < 1 Then Exit Sub
            For i As Integer = 0 To ListPaths.Items.Count - 1
                If ListPaths.GetItemChecked(i) = True Then
                    strItem = ListPaths.Items(i).ToString
                    Dim intRefId As Integer = FindRefId(strItem)
                    If intRefId > 0 Then
                        Dim k As Integer
                        k = DoAssignRefToProduct(intRefId, intProd)
                        If k = 1 Then
                            lblPath.Text = "Assigned:  " & strItem
                            tmpCNT = tmpCNT + 1
                        End If
                    End If
                End If
            Next i
            lblPath.Text = "Assigned Refs:  " & tmpCNT.ToString
        End If
    End Sub
    Private Function FindRefId(strTitle As String) As Integer
        '//Find ID of a Ref in tblPapers
        DS.Tables("tblRefs1").Clear()
        Dim Fltr As String = "PaperName='" & strTitle & "'"
        Using CnnSS = New SqlClient.SqlConnection(strDatabaseCNNstring)
            CnnSS.Open()
            DASS = New SqlClient.SqlDataAdapter("SELECT Distinct Papers.ID, PaperName, IsPaper, IsBook, IsManual, IsLecture, Papers.Note FROM [Paper_Product] RIGHT JOIN Papers ON [Paper_Product].Paper_ID = Papers.ID  WHERE (" + Fltr + ") ORDER BY Papers.ID DESC;", CnnSS)
            DASS.Fill(DS, "tblRefs1")
            CnnSS.Close()
        End Using
        FindRefId = DS.Tables("tblRefs1").Rows(0).Item(0)
    End Function
    Private Function DoAssignRefToProduct(iRef As Integer, iProd As Integer) As Integer
        strSQL = "INSERT INTO Paper_Product (Paper_ID, Product_ID) VALUES (@paperid, @productid)"
        DoAssignRefToProduct = 1
        Using CnnSS = New SqlClient.SqlConnection(strDatabaseCNNstring)
            CnnSS.Open()
            Dim cmdx As New SqlClient.SqlCommand(strSQL, CnnSS)
            cmdx.CommandType = CommandType.Text
            cmdx.Parameters.AddWithValue("@paperid", iRef.ToString)
            cmdx.Parameters.AddWithValue("@productid", iProd.ToString)
            Try
                Dim ix As Integer = cmdx.ExecuteNonQuery()
            Catch ex As Exception
                DoAssignRefToProduct = 0
                MsgBox(ex.ToString)
            End Try
            CnnSS.Close()
        End Using
    End Function
    Private Sub Menu2_LastVisited_Click(sender As Object, e As EventArgs) Handles Menu2_LastVisited.Click
        ScanFolder(DestinationFolder)
    End Sub
    Private Sub Menu2_Papers_Click(sender As Object, e As EventArgs) Handles Menu2_Papers.Click
        DestinationFolder = strFolderPapers & "\"
        ScanFolder(strFolderPapers)
    End Sub
    Private Sub Menu2_Books_Click(sender As Object, e As EventArgs) Handles Menu2_Books.Click
        DestinationFolder = strFolderBooks & "\"
        ScanFolder(strFolderBooks)
    End Sub
    Private Sub Menu2_Manuals_Click(sender As Object, e As EventArgs) Handles Menu2_Manuals.Click
        DestinationFolder = strFolderManuals & "\"
        ScanFolder(strFolderManuals)
    End Sub
    Private Sub Menu2_Lectures_Click(sender As Object, e As EventArgs) Handles Menu2_Lectures.Click
        DestinationFolder = strFolderLectures & "\"
        ScanFolder(strFolderLectures)
    End Sub
    Private Sub lblPath_Click(sender As Object, e As EventArgs) Handles lblPath.Click
        Menu_SelectFolder_Click(sender, e)
    End Sub
    Private Sub Menu_Exit_Click(sender As Object, e As EventArgs) Handles Menu_Exit.Click
        Me.Dispose()
    End Sub


End Class